src/github.com/getlantern/fronted/cache.go

```diff
@@ -0,0 +1,103 @@
+package fronted
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+const (
+	maxAllowedCachedAge = 24 * time.Hour
+	maxCacheSize        = 1000
+	cacheSaveInterval   = 5 * time.Second
+)
+
+var (
+	fillSentinel *Masquerade = nil
+)
+
+func (d *direct) initCaching() int {
+	prevetted := d.prepopulateMasquerades()
+	go d.fillCache()
+	return prevetted
+}
+
+func (d *direct) prepopulateMasquerades() int {
+	file, err := os.Open(d.cacheFile)
+	if err == nil {
+		log.Debugf("Attempting to prepopulate masquerades from cache")
+		defer file.Close()
+		var masquerades []*Masquerade
+		err := json.NewDecoder(file).Decode(&masquerades)
+		if err != nil {
+			log.Errorf("Error prepopulating cached masquerades: %v", err)
+			return 0
+		}
+
+		now := time.Now()
+		for _, m := range masquerades {
+			if now.Sub(m.LastVetted) < maxAllowedCachedAge {
+				select {
+				case d.masquerades <- m:
+					// submitted
+					d.cache = append(d.cache, m)
+				default:
+					// channel full, that's okay
+				}
+			}
+		}
+	}
+
+	return len(d.cache)
+}
+
+func (d *direct) fillCache() {
+	saveTimer := time.NewTimer(cacheSaveInterval)
+	cacheChanged := false
+	for {
+		select {
+		case m := <-d.toCache:
+			if m == fillSentinel {
+				log.Debug("Cache closed, stop filling")
+				return
+			}
+			log.Debugf("Caching vetted masquerade for %v (%v)", m.Domain, m.IpAddress)
+			d.cache = append(d.cache, m)
+			cacheChanged = true
+		case <-saveTimer.C:
+			if !cacheChanged {
```

Comment:
Thanks @oxtoacart!

On Thu, May 12, 2016 at 09:31 oxtoacart <notifications@github.com> wrote:

> In src/github.com/getlantern/fronted/cache.go
> <https://github.com/getlantern/lantern/pull/4272#discussion_r63030358>:
>
> > +
> > +func (d *direct) fillCache() {
> > +	saveTimer := time.NewTimer(cacheSaveInterval)
> > +	cacheChanged := false
> > +	for {
> > +		select {
> > +		case m := <-d.toCache:
> > +			if m == fillSentinel {
> > +				log.Debug("Cache closed, stop filling")
> > +				return
> > +			}
> > +			log.Debugf("Caching vetted masquerade for %v (%v)", m.Domain, m.IpAddress)
> > +			d.cache = append(d.cache, m)
> > +			cacheChanged = true
> > +		case <-saveTimer.C:
> > +			if !cacheChanged {
>
> @myleshorton <https://github.com/myleshorton> I took your suggestion and
> just write out the whole file everytime. For efficiency, I only write it if
> changed, and I still cap the total cache size just in case.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/pull/4272/files/07d4f1f637cc781d3969910763a391bde4ef0421#r63030358>
>